Document/0x05b-Basic-Security_Testing.md

```diff
@@ -366,7 +366,109 @@ More infomation on using the Objection REPL can be found on the [Objection Wiki]
+```
+
+```shell
+$ rafind2 -ZS service AndroidManifest.xml
+```
+
+```shell
+$ rafind2 -ZS receiver AndroidManifest.xml
+```
+
+You can use the flag `-a` to manually set the architecture to Dalvik.
+
+```shell
+$ r2 -a dalvik classes.dex
+```
+
+Once loaded, you can print the disassembly opcodes with the r2 command `pd`.
+
+```shell
+[0x000009c8]> pd
+```
+
+You can limit the number of opcodes displayed by appending the number to the command `pd`.
+
+```shell
+[0x000009c8]> pd 10
+            ;-- entry0:
+            ;-- Lsg/vantagepoint/uncrackable1/MainActivity.method.onCreate(Landroid/os/Bundle;)V:
+            ;-- method.protected.Lsg_vantagepoint_uncrackable1_MainActivity.Lsg_vantagepoint_uncrackable1_MainActivity.method.onCreate_Landroid_os_Bundle__V:
+            ;-- ip:
+            0x000009c8      710027000000   invoke-static {}, Lsg/vantagepoint/a/c.a()Z ; 0x27
+            0x000009ce      0a00           move-result v0
+        ,=< 0x000009d0      39000e00       if-nez v0, 0x000009ec
+        |   0x000009d4      710028000000   invoke-static {}, Lsg/vantagepoint/a/c.b()Z
+        ...
+            ; 0x29
+       ||   0x000009e6      0a00           move-result v0
+      ,===< 0x000009e8      38000700       if-eqz v0, 0x000009f6
+      |``-> 0x000009ec      1a003f00       const-string v0, str.Root_detected ; 0x11bf
+```
+
+You can also print the class names with the r2 command `icq`.
+
+```shell
+[0x000009c8]> icq
+0x0000069c [0x0000078c - 0x000007b8] Lsg/vantagepoint/a/a Ljava/lang/Object;
+0x000006bc [0x000007c8 - 0x000007ec] Lsg/vantagepoint/a/b Ljava/lang/Object;
+...
+0x0000073c [0x00000958 - 0x00000abc] Lsg/vantagepoint/uncrackable1/MainActivity Landroid/app/Activity;
+0x0000075c [0x00000acc - 0x00000bb2] Lsg/vantagepoint/uncrackable1/a Ljava/lang/Object;
+```
+
+You can also print the external methods with the r2 command `iiq`.
+
+```shell
+[0x000009c8]> iiq
+Landroid/app/Activity.method.<init>()V
+Landroid/app/Activity.method.onCreate(Landroid/os/Bundle;)V
+...
+0x0000075c [0x00000acc - 0x00000bb2]    230 class 6 Lsg/vantagepoint/uncrackable1/a super: Ljava/lang/Object;
+0x00000acc method 0 sp   Lsg/vantagepoint/uncrackable1/a.method.a(Ljava/lang/String;)Z
+0x00000b5c method 1 sp   Lsg/vantagepoint/uncrackable1/a.method.b(Ljava/lang/String;)[B
+```
+
+You can also print all the strings with the r2 command `izq`.
+
+```shell
+[0x000009c8]> izq
+0xc50 39 39 /dev/com.koushikdutta.superuser.daemon/
+0xc79 25 25 /system/app/Superuser.apk
+...
+0xd23 44 44 5UJiFctbmgbDoLXmpL12mkno8HT4Lv8dlat8FxR2GOc=
+0xd51 32 32 8d127684cbc37c17616d806cf50473cc
+0xd76 6 6 <init>
+0xd83 10 10 AES error:
+0xd8f 20 20 AES/ECB/PKCS7Padding
+0xda5 18 18 App is debuggable!
+0xdc0 9 9 CodeCheck
+0x11ac 7 7 Nope...
+0x11bf 14 14 Root detected!
+```
+
+Radare2 is very powerful and has dozens commands that you can find on the [radare2 command documentation](https://radare.gitbooks.io/radare2book/basic_commands/intro.html "radare2 command documentation").
```

Comment:
Not related specifically to this line:

Also mention that you can also use it for JNI

Reply:
Added it to the beginning of the section.